File: db_connector.py
import mysql.connector
from typing import Dict, Any
from util import normalize_sql_definition

class DBConnector:
    def __init__(self):
        self.connection = None
        
    def connect(self, config: Dict[str, Any]) -> None:
        """连接到MySQL数据库"""
        try:
            self.connection = mysql.connector.connect(
                host=config['host'],
                port=config['port'],
                user=config['user'],
                password=config['password'],
                database=config['database']
            )
        except Exception as e:
            raise Exception(f"连接数据库失败: {str(e)}")
            
    def get_table_structure(self) -> Dict[str, Dict[str, Any]]:
        """获取所有表的结构"""
        if not self.connection:
            raise Exception("未连接到数据库")
            
        cursor = self.connection.cursor(dictionary=True)
        tables = {}
        
        try:
            # 获取所有表名
            cursor.execute("SHOW TABLES")
            table_names = [row[f'Tables_in_{self.connection.database}'] for row in cursor.fetchall()]
            
            for table_name in table_names:
                # 获取表结构
                cursor.execute(f"SHOW CREATE TABLE `{table_name}`")
                create_table_sql = cursor.fetchone()['Create Table']
                
                # 获取列信息
                cursor.execute(f"SHOW FULL COLUMNS FROM `{table_name}`")
                columns = {}
                for col in cursor.fetchall():
                    col_name = col['Field']
                    col_type = col['Type']
                    col_null = 'NULL' if col['Null'] == 'YES' else 'NOT NULL'
                    col_default = f"DEFAULT {col['Default']}" if col['Default'] is not None else ''
                    col_extra = col['Extra']
                    col_comment = f"COMMENT '{col['Comment']}'" if col['Comment'] else ''
                    
                    # 组合列定义 
                    # {'Field': 'team_id', 'Type': 'char(32)', 'Collation': 'utf8mb4_0900_ai_ci', 'Null': 'NO', 'Key': '', 'Default': '', 'Extra': '', 'Privileges': 'select,insert,update,references', 'Comment': '组织ID'}
                    col_def = f"{col_type} {col_null} {col_default} {col_extra} {col_comment}".strip()
                    columns[col_name] = {
                        'raw': col_def,
                        'normalized': normalize_sql_definition(col_def),
                        'details': {
                            "Type": col['Type'],
                            "Collation": col['Collation'],
                            "Null": col['Null'],
                            "Key": col['Key'],
                            "Default": col['Default'],
                            "Extra": col['Extra'],
                            "Comment": col['Comment'],
                        }
                    }
                
                # 获取索引信息
                cursor.execute(f"SHOW INDEX FROM `{table_name}`")
                indexes = {}
                for idx in cursor.fetchall():
                    idx_name = idx['Key_name']
                    if idx_name == 'PRIMARY':
                        idx_type = 'PRIMARY KEY'
                    elif idx['Non_unique'] == 0:
                        idx_type = 'UNIQUE KEY'
                    else:
                        idx_type = 'KEY'
                        
                    if idx_name not in indexes:
                        indexes[idx_name] = {
                            'type': idx_type,
                            'columns': idx['Column_name']
                        }
                    else:
                        # 如果索引包含多个列，将它们组合起来
                        indexes[idx_name]['columns'] += f", {idx['Column_name']}"
                
                tables[table_name] = {
                    'columns': columns,
                    'indexes': indexes,
                    'raw_sql': create_table_sql
                }
                
        finally:
            cursor.close()
            
        return tables
        
    def close(self):
        """关闭数据库连接"""
        if self.connection:
            self.connection.close()
            self.connection = None 

File: src/main/db-connector.ts
import mysql from 'mysql2/promise';
import { Tables } from './types';

export class DBConnector {
  private connection: mysql.Connection | null = null;

  async connect(config: any): Promise<void> {
    try {
      this.connection = await mysql.createConnection({
        host: config.host,
        port: config.port || 3306,
        user: config.username || config.user,
        password: config.password,
        database: config.database,
        charset: 'utf8mb4'
      });
    } catch (error) {
      throw new Error(`连接数据库失败: ${error}`);
    }
  }

  async close(): Promise<void> {
    if (this.connection) {
      await this.connection.end();
      this.connection = null;
    }
  }

  async getTableStructure(): Promise<Tables> {
    if (!this.connection) {
      throw new Error('数据库未连接');
    }

    const tables: Tables = {};

    try {
      // 获取所有表名
      const [tableRows] = await this.connection.execute(
        "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA = ?",
        [this.connection.config.database]
      ) as any[];

      for (const tableRow of tableRows) {
        const tableName = tableRow.TABLE_NAME;
        const tableStructure = await this.getTableStructureByName(tableName);
        tables[tableName] = tableStructure;
      }
    } catch (error) {
      throw new Error(`获取表结构失败: ${error}`);
    }

    return tables;
  }

  private async getTableStructureByName(tableName: string): Promise<any> {
    if (!this.connection) {
      throw new Error('数据库未连接');
    }

    const columns: { [key: string]: any } = {};
    const indexes: { [key: string]: any } = {};

    try {
      // 获取列信息
      const [columnRows] = await this.connection.execute(`
        SELECT 
          COLUMN_NAME,
          DATA_TYPE,
          CHARACTER_MAXIMUM_LENGTH,
          IS_NULLABLE,
          COLUMN_DEFAULT,
          COLUMN_COMMENT
        FROM INFORMATION_SCHEMA.COLUMNS 
        WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
        ORDER BY ORDINAL_POSITION
      `, [this.connection.config.database, tableName]) as any[];

      for (const columnRow of columnRows) {
        const columnName = columnRow.COLUMN_NAME;
        const dataType = columnRow.DATA_TYPE;
        const maxLength = columnRow.CHARACTER_MAXIMUM_LENGTH;
        const isNullable = columnRow.IS_NULLABLE === 'YES';
        const defaultValue = columnRow.COLUMN_DEFAULT;
        const comment = columnRow.COLUMN_COMMENT;

        let typeDef = dataType;
        if (maxLength) {
          typeDef += `(${maxLength})`;
        }

        let rawDef = `\`${columnName}\` ${typeDef}`;
        if (!isNullable) {
          rawDef += ' NOT NULL';
        }
        if (defaultValue !== null) {
          rawDef += ` DEFAULT ${defaultValue}`;
        }
        if (comment) {
          rawDef += ` COMMENT '${comment}'`;
        }

        columns[columnName] = {
          raw: rawDef,
          type: dataType,
          length: maxLength ? maxLength.toString() : undefined,
          nullable: isNullable,
          default: defaultValue !== null ? `DEFAULT ${defaultValue}` : undefined,
          comment: comment ? `COMMENT '${comment}'` : undefined
        };
      }

      // 获取索引信息
      const [indexRows] = await this.connection.execute(`
        SELECT 
          INDEX_NAME,
          NON_UNIQUE,
          GROUP_CONCAT(COLUMN_NAME ORDER BY SEQ_IN_INDEX) as COLUMNS
        FROM INFORMATION_SCHEMA.STATISTICS 
        WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
        GROUP BY INDEX_NAME, NON_UNIQUE
        ORDER BY INDEX_NAME
      `, [this.connection.config.database, tableName]) as any[];

      for (const indexRow of indexRows) {
        const indexName = indexRow.INDEX_NAME;
        const isUnique = !indexRow.NON_UNIQUE;
        const indexColumns = indexRow.COLUMNS;

        // 跳过主键
        if (indexName === 'PRIMARY') {
          continue;
        }

        indexes[indexName] = {
          type: isUnique ? 'UNIQUE' : 'INDEX',
          columns: indexColumns,
          unique: isUnique
        };
      }
    } catch (error) {
      throw new Error(`获取表 ${tableName} 结构失败: ${error}`);
    }

    return { columns, indexes };
  }
} 